File: 总题库/21.MergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if l1 is None:
            return l2
        if l2 is None:
            return l1

        dummy = ListNode(0)
        preNode = dummy
        while l1 != None and l2 != None:
            if l1.val < l2.val:
                preNode.next = l1
                l1 = l1.next
            else:
                preNode.next = l2
                l2 = l2.next
            preNode = preNode.next
        
        while l1 is not None:
            preNode.next = l1
            l1 = l1.next
            preNode = preNode.next

        while l2 is not None:
            preNode.next = l2
            l2 = l2.next
            preNode = preNode.next
        return dummy.next

File: 总题库/21.MergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// 使用哑节点可以使用循环
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == NULL) return l2;
        if (l2 == NULL) return l1;
        ListNode dummy(0);
        ListNode *iter = &dummy;
        while (l1 != NULL && l2 != NULL) {
            if (l1->val < l2->val) {
                iter->next = l1;
                l1 = l1->next;
            }
            else {
                iter->next = l2;
                l2 = l2->next;
            }
            iter = iter->next;
        }
        while (l1 != NULL) {
            iter->next = l1;
            l1 = l1->next;
            iter = iter->next;
        }
        while (l2 != NULL) {
            iter->next = l2;
            l2 = l2->next;
            iter = iter->next;
        }
        return dummy.next;
    }
};